Application Default Credentials after API polishing

jtattermusch opened the pull request:
This PR contains code from #553, with the latest changes from upstream/master merged.
I originally created this branch just to facilitate cleaning up the git history and to fix some of the style nits in the code.

Nevertheless, once I started editing the code, I realized there were some major issues with the API in the original pull request (perhaps introduced while addressing review comments)  - for example the object returned from GoogleAuth.GetApplicationDefaultCredentials() didn't expose the IsCreateScopedRequired property and accessing that property required guessing that the default credential is a service account credential in fact, explicitly typecasting it and only after that, you were able to access the IsCreateScopedRequired property - which makes no sense at all. I found other problems as well (those that are hard to catch in review until you actually start modifying the code).

So, because there were some concerns about the API in the past, i decided to take the app default credentials code (I haven't changed most of the code that implements the actual logic) and adjust the API layout.

The new API introduces a new type GoogleCredential (returned by GetApplicationDefaultCredentials()), which acts as a very thin wrapper for other credential types. That way, the API of the existing credential types could stay intact and GoogleCredential can still provide all the functionality needed. Arguably, the new design is cleaner and possibly allows more tweaking in the future if necessary.

The code is still not 100% finished, but I believe it is in good shape for an initial review (of the API and high level changes I've made).

@vsubramani  @peleyal  for review

vsubramani commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,280 @@
+        /// <summary>
+        /// Wraps <c>ServiceAccountCredential</c> as <c>GoogleCredential</c>.
+        /// </summary>
+        internal class ServiceAccountGoogleCredential : GoogleCredential
```
It isn't clear why the ServiceAccountGoogleCredential type is needed. Why not use the ServiceAccountCredential itself? Same for the UserAccountGoogleCredential and ComputeGoogleCredential.

As a library consumer, there now ends up being multiple similar credential types which might end up confusing the user. 

ServiceAccountCredential
ServiceAccountGoogleCredential
UserCredential
UserGoogleCredential
ComputeCredential
ComputeGoogleCredential

jtattermusch replied:
The wrapper classes like ServiceAccountGoogleCredential, UserGoogleCredential and ComputeGoogleCredential are internal and user will never see them. They are used just to adapt the API appropriately.

vsubramani replied:
I missed that part initially. Thanks for clarifying.

However, the credential.UnderlyingCredential seems a bit odd usage pattern though.

jtattermusch replied:
UnderlyingCredential is also internal. I use it just for testing.

vsubramani replied:
LGTM on the changes. This is close to the API surface that we pushed for originally.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,232 @@
+
+        // TODO(jtattermusch): test compute engine credential.
+
+        // TODO(jtattermusch): add more tests
```
Which tests do you want to add? Please elaborate  more.

jtattermusch replied:
This was basically a note to self. I added a few more tests and removed the TODO.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,232 @@
+            try
+            {
+                var credential = credentialProvider.GetApplicationDefaultCredential();
+                Assert.Fail();
```
Please add a massage here, explaining what went wrong.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,232 @@
+
+            var scopes = new[] { "https://www.googleapis.com/auth/cloud-platform" };
+            var scopedCredential = credential.CreateScoped(scopes);
+            Assert.IsTrue(credential != scopedCredential);
```
Assert.NotEquals or something like that?

jtattermusch replied:
changed to !object.ReferenceEquals.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,40 @@
+// The assembly version has the format "{Major}.{Minor}.{Build}.{Revision}".
+// The form "{Major}.{Minor}.*" will automatically update the build and revision,
+// and "{Major}.{Minor}.{Build}.*" will update just the revision.
+[assembly: AssemblyVersion("1.9.1.*")]
```
Need to update now :)
1.9.2 :)

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+using Google.Apis.Http;
+using Google.Apis.Logging;
+using Google.Apis.Json;
+using Google.Apis.Util;
```
A nit :)
Out of order. Please fix

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        private bool checkedComputeEngine = false;
+
+        /// <summary>
+        /// Returns the Application Default Credentials
```
"." in the end of sentence.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        /// </summary>
+        public GoogleCredential GetApplicationDefaultCredential()
+        {
+            if (cachedCredential == null)
```
Optional - link to https://msdn.microsoft.com/en-us/library/ff650316.aspx?

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        private const string CloudSDKConfigDirectory = "gcloud";
+
+        /// <summary>Help link to the application default credentials feature.</summary>
+        private const string HelpPermalink = "https://developers.google.com/accounts/docs/application-default-credentials";
```
Since it is only used once, should we just include it in the exception?
What do you think?

jtattermusch replied:
I don't think that's necessary, but I can address that in the very last phase of review if you insist.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+            }
+            catch (Exception e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
```
Which line from 104 to 108 can throw the exception? Let's limit the try and catch only to the throwing function.
Same below...

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+            }
+            catch (FileNotFoundException)
+            {
+                // File is not present, eat the exception and move on to the next check.
```
Log?

jtattermusch replied:
I think this should remain silent as when running on GCE you would always see a warning that the well known file hasn't been found  - but on GCE it's not expected to be there, so this would be misleading.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+            if (!checkedComputeEngine)
+            {
+                Logger.Debug("Checking whether the application is running on ComputeEngine.");
+                bool isRunningOnComputeEngine = ComputeCredential.IsRunningOnComputeEngine().Result;
```
Can we avoid the Result here?
Can we use async and await?

jtattermusch replied:
Switched to async where needed.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+                if (isRunningOnComputeEngine)
+                {
+                    Logger.Debug("ComputeEngine check passed. Loading ComputeEngine Credentials.");
+                    return GoogleCredential.FromCredential(new ComputeCredential());
```
Seems weird to me. How does it help to create an empty object and then send it to GoogleCredential. Maybe I need more context, but it seems weird.. What do you think?

jtattermusch replied:
new ComputeCredential() creates a full-fledged compute credential. For computeCredential you don't really need any initialization parameter, it works as such. I need to wrap it here so it has the interface I need.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+                }
+            };
+
+            var token = new TokenResponse()
```
nit, you can remove the () here.

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        }
+
+        /// <summary>Creates default credential from a JSON file.</summary>
+        internal GoogleCredential CreateDefaultCredentialFromFile(string credentialPath)
```
Are methods here internal in purpose? Is it for testing purposes? If so, I would add the https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Core/Apis/Testing/VisibleForTestOnly.cs annotation

jtattermusch replied:
Marked everything I don't use in tests as private.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        /// </summary>
+        internal string GetWellKnownCredentialFilePath()
+        {
+            return Path.Combine(GetEnvironmentVariable(AppdataEnvironmentVariable) ?? "",
```
Does it make sense to have
"GetEnvironmentVariable(AppdataEnvironmentVariable) ?? ""," in a new line? It seems like more readable to me.
What do you think?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,252 @@
+        }
+
+        /// <summary>Helper method to get the environment variable. Can be overriden for testing.</summary>
+        protected virtual string GetEnvironmentVariable(string variableName)
```
I would rephrase it. something like:
Gets the environment variable. This method is protected so it could be overriden for testing purposes only.

How does that sound? Same change to line 246?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleAuth.cs:
```diff
@@ -0,0 +1,66 @@
+namespace Google.Apis.Auth.OAuth2
+{
+    /// <summary>Provides the Application Default Credential from the environment.</summary>
+    public static class GoogleAuth
```
I'm still not convinced why do we need this wrapper. Can you elaborate...

jtattermusch replied:
I think we suffciently discussed that offline.
Now this class is gone and the static method has moved to GoogleCredential.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleAuth.cs:
```diff
@@ -0,0 +1,66 @@
+        /// </summary>
+        public static GoogleCredential GetApplicationDefaultCredential()
+        {
+            return defaultCredentialProvider.GetApplicationDefaultCredential();
```
If we leave this class as it is, which I'm still not convinced.
Does it worth to change the content of GetApplicationDefaultCredential to be in the constructor and then we don't need the double singleton check. We can just create it statically and that's all :)

peleyal replied:
I'm not sure that I like to do all the calls that you have in the constructor. It just me thinking out loudly :)

jtattermusch replied:
This is getting refactored a bit so let's see if you will like the resulting version more.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -99,5 +116,48 @@ public Initializer(string tokenUrl)
+                CancellationTokenSource cts = new CancellationTokenSource();
+                cts.CancelAfter(MetadataServerPingTimeoutInMilliseconds);
+
+                // Using the built-in HttpClient, as we want bare bones functionality without any retries.
```
good comment

peleyal commented on Tools/Google.Apis.Release/Resources/Google.Apis.Auth.VERSION.nuspec:
```diff
@@ -25,8 +25,8 @@
         </description>
         <tags>Google</tags>
         <dependencies>
-          <dependency id="BouncyCastle" version="1.7.0" />
```
revert?

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,280 @@
+
+            internal override object UnderlyingCredential
+            {
+                get
```
nit - just a style improvement:
get { return credential; } ONE LINE :)

jtattermusch replied:
done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,280 @@
+    /// credentials (like ServiceAccountCredential, ComputeCredential or UserCredential)
+    /// in a unified way.
+    /// </summary>
+    public abstract class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
```
Why did you decide to use this pattern. Should we create a factory class instead?
I'm still going through this class, trying to understand if it's the right design...

peleyal commented:
I need to put more thoughts in the GoogleCredential usage here and the GoogleAuth classes.
Meantime that's what I have for you.

I'll try to take another look tomorrow as well
Thanks

jtattermusch commented:
Ok, so I think I addressed most of the comments. The list of important changes I made:

-- I added caching using that uses Lazy directly to IsRunningOnComputeEngine
-- The most important entry point is now static method    GoogleCredential.GetApplicationDefaultAsync(). I also exposed GoogleCredential.FromStream (that can read credential from JSON data).
-- I made the original GetApplicationDefaultCredential() method async. That allows me to call async methods in the implementation cleanly and mitigates risk for running into deadlock (using .Result in library code can be pretty dangerous). We can reconsider adding the sync method later perhaps.
-- I changed the way application default credentials are cached (I am also using Lazy<> here)
-- I didn't get rid of DefaultCredentialProvider as I originally wanted, because having a separate class like this makes testing easier and without it I would have to come up with some ugly hacks. This class contains a pretty well defined functionality and it's internal, so I don't think this should be a problem.
-- I added some tests

@peleyal  please take a look and add more comments as needed. My feeling is that unless I missed something, the current code should be pretty solid and fine for shipping.

@vsubramani  please take a look at the API changes I made, I don't think they are controversial, but I definitely wanted to check with you.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/GoogleApis.Auth.DotNet4.Tests.csproj:
```diff
@@ -0,0 +1,141 @@
+      <Project>{db18f3a6-b16a-477b-a7b1-c2b116d1bfb0}</Project>
+      <Name>GoogleApis.Core</Name>
+    </ProjectReference>
+    <ProjectReference Include="..\GoogleApis.Tests\GoogleApis.Tests.csproj">
```
Why do you need it?

jtattermusch replied:
The reference to GoogleApis.Tests? I probably don't, good point.

peleyal replied:
yep GoogleApis.Tests. Remove if you can...

jtattermusch replied:
Removed.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/packages.config:
```diff
@@ -0,0 +1,10 @@
+<packages>
+  <package id="Microsoft.Bcl" version="1.1.10" targetFramework="net40" />
+  <package id="Microsoft.Bcl.Async" version="1.0.168" targetFramework="net40" />
+  <package id="Microsoft.Bcl.Build" version="1.0.14" targetFramework="net40" />
```
Please upgrade to 1.0.21

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    /// <summary>A mock for the <see cref="Google.Apis.Auth.OAuth2.DefaultCredentialProvider"/>.</summary>
+    class TestDefaultCredentialProvider : DefaultCredentialProvider
```
s/TestDefaultCredentialProvider/MockDefaultCredentialProvider or FakeDefaultCredentialProvider

peleyal replied:
I would put it at the beginning of the file, just because I did in other tests files as well...

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,271 @@
+        [Test]
+        public async Task GetDefaultCredential_UserCredential_FromEnvironmentVariable()
+        {
+            string credentialFilepath = "TempFilePath.json";
```
s/string/var

jtattermusch replied:
Done here and other places in this file.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,271 @@
+            try
+            {
+                var credential = await credentialProvider.GetDefaultCredentialAsync();
+                Assert.Fail();
```
Should we add a message here?

jtattermusch replied:
I don't think so. This is a common pattern and adding the message wouldn't really increase the clarity, it would just bloat the code.

peleyal replied:
I think that it will make it easier to read \ understand the output when you run several tests.
I would recommend you adding it, but it's up to you.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/GoogleCredentialTests.cs:
```diff
@@ -0,0 +1,92 @@
+using System;
+using System.Threading.Tasks;
+
+using NUnit.Framework;
```
empty line after NUnit

jtattermusch replied:
done.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/GoogleCredentialTests.cs:
```diff
@@ -0,0 +1,92 @@
+                    credential = credential.CreateScoped(new[] { "the_scope" });
+                }
+            }
+            catch (InvalidOperationException)
```
??? Is it expected?

jtattermusch replied:
I only added this test to demonstrate the usage pattern. Because no app default credentials are set in the test, the snippet will fail.  If you think that's misleading, I will just remove this "test".

jtattermusch replied:
removed.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+        private static readonly string CloudSDKConfigDirectoryUnix = Path.Combine(".config", "gcloud");
+
+        /// <summary>Caches result from first call to <c>GetApplicationDefaultCredentialAsync</c> </summary>
+        private readonly Lazy<Task<GoogleCredential>> cachedCredentialTask;
```
Nice!

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+        /// <summary>Caches result from first call to <c>GetApplicationDefaultCredentialAsync</c> </summary>
+        private readonly Lazy<Task<GoogleCredential>> cachedCredentialTask;
+
+        public DefaultCredentialProvider()
```
Please add a comment, even if it's a stupid one :)
/// <summary>
/// Constructs a new default credential provider.
/// </summary>

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+
+        public DefaultCredentialProvider()
+        {
+            this.cachedCredentialTask = new Lazy<Task<GoogleCredential>>(CreateDefaultCredentialAsync);
```
Can we remove this?

jtattermusch replied:
Not sure what you mean.

peleyal replied:
Can we remove "this" I don't think that you need it.
cachedCredentialTask = new Lazy<Task<GoogleCredential>>(CreateDefaultCredentialAsync);

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+                {
+                    // Catching generic exception type because any corrupted file could manifest in different ways including 
+                    // but not limited to the System, System.IO or from the Newtonsoft.Json namespace.
+
```
remove empty line

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+                }
+                catch (FileNotFoundException)
+                {
+                    // File is not present, eat the exception and move on to the next check.
```
Log?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,271 @@
+            credentialProvider.SetEnvironmentVariable(CredentialEnvironmentVariable, credentialFilepath);
+            credentialProvider.SetFileContents(credentialFilepath, DummyUserCredentialFileContents);
+
+            var credential = await credentialProvider.GetDefaultCredentialAsync();
```
ConfigureAwait(false) ?

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,271 @@
+            credentialProvider.SetEnvironmentVariable("APPDATA", AppDataValue);
+            credentialProvider.SetFileContents(WellKnownCredentialFilePath, DummyUserCredentialFileContents);
+
+            var credential = await credentialProvider.GetDefaultCredentialAsync();
```
ConfigureAwait(false) ?

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+        /// <summary>Creates a new default credential.</summary>
+        private async Task<GoogleCredential> CreateDefaultCredentialAsync()
+        {
+            // 1. First try the environment variable.
```
I think that the order of this code should be documented somewhere, maybe in the class level.
I mean the order of

// 1. First try the environment variable.
// 2. Then try the well known file.
...

jtattermusch replied:
The list of steps is documented in detail on GoogleCredentials.GetApplicationDefaultAsync() which is the publicly exposed API.

peleyal replied:
Sweet. That's enough

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+            return CreateDefaultCredentialFromJson(credentialParameters);
+        }
+
+        /// <summary>Creates a default credential from JSON data. </summary>
```
Please remove space after "."
/// <summary>Creates a default credential from JSON data.</summary>

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+                RefreshToken = credentialParameters.RefreshToken
+            };
+
+            var flow = new GoogleAuthorizationCodeFlow(initializer);
```
I would change the order of the code.
var token... should be before the initialzier.
initializer should be close to the place that it's being used...

jtattermusch replied:
done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,258 @@
+            }
+        }
+
+        private static UserCredential CreateUserCredentialFromJson(JsonCredentialParameters credentialParameters)
```
Can you add a comment to private method as well?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Http;
+using System.IO;
```
please fix the order

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+    /// <summary>
+    /// Credential for authorizing calls using OAuth 2.0.
+    /// It is a convenience wrapper that allows handling of different types of 
+    /// credentials (like ServiceAccountCredential, ComputeCredential or UserCredential)
```
Do you want to add a link to ServiceAccountCredential, ComputeCredential or UserCredential?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+    /// </summary>
+    public abstract class GoogleCredential : IConfigurableHttpClientInitializer, ITokenAccess
+    {
+        private static DefaultCredentialProvider defaultCredentialProvider = new DefaultCredentialProvider();
```
Why GoogleCredentail need to have a reference to DefaultProvider?
I didn't expect it... what am I missing?

jtattermusch replied:
default provider  has the implementation  of GetApplicationDefaultAsync(). As we discussed before, it is useful to have default provider as a separate class because it makes testing easier.

jtattermusch replied:
Added a TODO to DefaultCredentialProvider, as discussed offline.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+        /// Console or a stored user credential using the format supported by the Cloud SDK.
+        /// </para>
+        /// </summary>
+        public static GoogleCredential FromStream(Stream stream)
```
Why do we need to expose this method? I thought that for the outside world we want to include only "GetApplicationDefaultAsync"

jtattermusch replied:
This actually comes handy because it allows you to read JSON service account credential  or JSON user credential from a file explicitly. 
Imagine you download the JSON key file from cloud console and want to use it with your app. If you used ServiceAccountCredential directly, you would need to parse the JSON first and extract values of the fields and initialize service account credential accordingly. 
This method can do all of that.

peleyal replied:
ACK.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+        // to the following protected abstract members.
+        protected abstract void Initialize(ConfigurableHttpClient httpClient);
+
+        protected abstract TokenResponse Token { get; }
```
Please add comments.

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+
+        #region Factory methods
+
+        internal static GoogleCredential FromCredential(ComputeCredential credential)
```
ditto. comments

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,299 @@
+        /// </summary>
+        internal class ComputeGoogleCredential : GoogleCredential
+        {
+            private readonly ComputeCredential credential;
```
TODO:
(or you can have it)

I think that any Credential should inehrit from some ICredential interface that includes the Initialize (already from IConfigurableHttpClientInitializer), Token and RequestAccessTokenAsync.
then you can avoid all the following methods for ComputeGoogleCredential, ServiceAccountGoogleCredential and UserGoogleCredential

peleyal replied:
Cool, you already have it  ITokenAccess, so maybe we can change my classes as well.
TODO. Not urgent. But it will make your code easier, right?

Or that's what we are trying to avoid here? Not changing my classes at all?

jtattermusch replied:
Yes, I see your point, but let's iterate. This is some 20 additional lines of code that are not visible externally at all and they allows us to do the job without messing with the existing API. Once this PR is in, introducing a common interface as you suggest would be a simple PR.

jtattermusch replied:
Added a TODO for this, as discussed offline.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/ITokenAccess.cs:
```diff
@@ -0,0 +1,40 @@
+    /// Provides members to retrieve auth token out-of-band.
+    /// This is necessary for workflows where you don't want to use 
+    /// <see cref="T:Google.Apis.Services.BaseClientService"/> to access the API.
+    /// If you don't know what this interface is for, chances are you don't need it.
```
Please remove "If you don't know what this interface is for, chances are you don't need it."

Provides members to retrieve auth token out-of-band.  --- CAN YOU ELABORE?
This is necessary for workflows where you don't want to use Services.BaseClientService to access the API. -- WHY???

jtattermusch replied:
because gRPC :-) Explained in the OO design document.
Will fix the grammar in the comment.

jtattermusch replied:
Rephrased the comment and added more clarification.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/JsonCredentialParameters.cs:
```diff
@@ -0,0 +1,61 @@
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Json;
+using Google.Apis.Util;
+using Google.Apis.Auth.OAuth2.Responses;
```
please resort

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.Tests/OAuth2/ComputeCredentialTests.cs:
```diff
@@ -0,0 +1,40 @@
+﻿/*
+Copyright 2013 Google Inc
```
2015

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -36,6 +38,25 @@ namespace Google.Apis.Auth.OAuth2
+        private const string MetadataFlavor = "Metadata-Flavor";
+
+        /// <summary>The Metadata header response indicating Google.</summary>
+        private const string Google = "Google";
```
rename to GoogleMetadataHeader?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -99,5 +120,53 @@ public Initializer(string tokenUrl)
+        {
+            try
+            {
+                Logger.Info("Checking connectivity to ComputeEngine metadata server.");
```
can we please limit the catch area to the ones that can actually throw exception?

jtattermusch replied:
I don't feel like this would be good for readability, in this case it's pretty obvious where does the exception come from and we only catching well-constrained exception types.

peleyal replied:
Is it only the SendAsync?

jtattermusch replied:
I think so.

peleyal replied:
In this case I prefer to try catch only it.
Up to you.... I'm just telling you that it will be easier to understand.

jtattermusch replied:
Let's leave it like this.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -99,5 +120,53 @@ public Initializer(string tokenUrl)
+            {
+                Logger.Info("Checking connectivity to ComputeEngine metadata server.");
+                var httpRequest = new HttpRequestMessage(HttpMethod.Get, MetadataServerUrl);
+                CancellationTokenSource cts = new CancellationTokenSource();
```
var?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -99,5 +120,53 @@ public Initializer(string tokenUrl)
+                // Using the built-in HttpClient, as we want bare bones functionality without any retries.
+                var httpClient = new HttpClient();
+                var response = await httpClient.SendAsync(httpRequest, cts.Token).ConfigureAwait(false);
+                cts.CancelAfter(MetadataServerPingTimeoutInMilliseconds);
```
shouldn't this line be before line 144????

jtattermusch replied:
This one looks safer, initializing Http client for the first time might take some time and our timeout is not very long.

jtattermusch replied:
Ah, my bad.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,270 @@
+        [Test]
+        public async Task GetDefaultCredential_UserCredential_FromEnvironmentVariable()
+        {
+            var credentialFilepath = "TempFilePath.json";
```
Mmmm... is it a file that we need to add to the project? It's unclear to me...

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,270 @@
+
+            Assert.IsInstanceOf(typeof(UserCredential), credential.UnderlyingCredential);
+            Assert.IsFalse(credential.IsCreateScopedRequired);
+            Assert.AreSame(credential, credential.CreateScoped(new [] { "scope1" }));
```
and where scope1 comes from?

peleyal replied:
ping...

jtattermusch replied:
Changed.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,270 @@
+        #region UserCredential
+
+        [Test]
+        public async Task GetDefaultCredential_UserCredential_FromEnvironmentVariable()
```
In general, I would add comments to the tests, explaining what are we testing in each kind. You can have it in the test level or before the Assert lines. Even for me that is reviewing this PR on and off, it's not that clear, why you have several of the following Asserts...
does it make sense?

jtattermusch replied:
I am leaning against it. Usually when you add comments to your test, you are just duplicating the information that is available in the test name already and that later becomes pain for maintenance (and you often end up with mismatch in test name and its comment after some period of time).

This test is called GetDefaultCredential_UserCredential_FromEnvironmentVariable,
so it tests the GetDefaultCredential method in the case where it should return UserCredential that is located based on an environment variable. I can still add a comment but I don't feel it would actually help the readability in the long run.

peleyal commented on Src/GoogleApis.Auth.DotNet4.Tests/OAuth2/DefaultCredentialProvider/DefaultCredentialProviderTests.cs:
```diff
@@ -0,0 +1,270 @@
+            var scopedCredential = credential.CreateScoped(scopes);
+            Assert.AreNotSame(credential, scopedCredential);
+
+            Assert.IsFalse(scopedCredential.IsCreateScopedRequired);
```
Please add Assert.IsInstanceOf(typeof(ServiceAccountCredential), credential.UnderlyingCredential); as well

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+            }
+
+            // 3. Then try the compute engine.     
+            Logger.Debug("Checking whether the application is running on ComputeEngine.");
```
Just curious, why did you add Logger.Debug here and not before Step 1 and 2?

jtattermusch replied:
Not sure if this is an excellent explanation, but it's because the check actually sends an http request to metadata server, which is something the user might wanna know about (previous cases only look for local files which is probably less important).

jtattermusch replied:
Actually, CreateDefaultCredentialFromFile method also logs a debug message. Sorry for not realizing this first.

peleyal replied:
ack

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+                return GoogleCredential.FromCredential(new ComputeCredential());
+            }
+
+            // If everything we could try has failed, report an error.
```
My English isn't great, so I'll leave it to you, but I would change it to:
// If everything we tried has failed, throw an exception.

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+        /// <summary>Creates default credential from a JSON file.</summary>
+        private GoogleCredential CreateDefaultCredentialFromFile(string credentialPath)
+        {
+            Logger.Debug(String.Format("Loading Credential from file {0}", credentialPath));
```
I just noticed that you don't need to use String.Format when using Logger.Debug (or other Logger method),
take a look at:
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis/Apis/%5BMedia%5D/Upload/ResumableUpload.cs#L527
and
https://github.com/google/google-api-dotnet-client/blob/master/Src/GoogleApis.Core/Apis/Logging/ILogger.cs

Can you please change it everywhere you used it?
Thanks

jtattermusch replied:
Yeah, thanks for noticing that. I found 2 locations where I was doing this and fixed them.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+            }
+        }
+
+        /// <summary>Creates default credential from a stream that contains JSON credential data.</summary>
```
Add "a"
Creates *a* default credential

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+            return new UserCredential(flow, "ApplicationDefaultCredentials", token);
+        }
+
+        /// <summary>Creates service account credential from JSON data.</summary>
```
Creates a service account...

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+                string.IsNullOrEmpty(credentialParameters.ClientEmail) ||
+                string.IsNullOrEmpty(credentialParameters.PrivateKey))
+            {
+                throw new InvalidOperationException("JSON data do not represent a valid service account credential.");
```
JSON data does not represent?

jtattermusch replied:
Actually "data" is plural in english (the reason for that being that the word "data" is plural in Latin originally, which sounds a bit strange, but indeed - In latin "data" basically means "things that have been given").

peleyal replied:
But you would say data is good (and not data are good).. right? data goes and NOT data go....

jtattermusch replied:
Let's not have conversation about something that seems to be the least important thing in this PR.
Changed to "does"

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+    /// <summary>
+    /// Credential for authorizing calls using OAuth 2.0.
+    /// It is a convenience wrapper that allows handling of different types of 
+    /// credentials (like <see cref="ServiceAccountCredential"/>, <see cref="ComputeCredential"/>
```
I'm not sure bout the following, please check:
Can you just verify if the link works, Do you need to include the full namespace? Or since it's in the same namespace you don't.

jtattermusch replied:
I think it should work if the name is in scope (you've imported it by using).
http://stackoverflow.com/questions/5994551/xml-comments-should-see-references-be-fully-qualified

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+    /// credentials (like <see cref="ServiceAccountCredential"/>, <see cref="ComputeCredential"/>
+    /// or <see cref="UserCredential"/>) in a unified way.
+    /// <para>
+    /// See <see cref="GetApplicationDefaultCredential"/>.
```
Improve the comment:
See ... for more details about the logic of getting the right credential?

jtattermusch replied:
Done and rephrased a bit.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <list type="number">
+        /// <item> 
+        /// <description>The environment variable GOOGLE_APPLICATION_CREDENTIALS is checked. If this variable is specified, it 
+        /// should point to a file that defines the credentials. The simplest way to get a credential for this purpose is to 
```
should print --> should print it ?

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <description>If you are running in Google Compute Engine production, the built-in service account associated with the 
+        /// virtual machine instance will be used.</description> 
+        /// </item> 
+        /// </list>
```
What happens if all the 3 items didn't match. What do we return than?

jtattermusch replied:
Added clarification.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        }
+
+        /// <summary>
+        /// Load credential from stream containing JSON credential data.
```
Load credetnial -> Loads the credential from a stream

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        }
+
+        /// <summary>
+        /// <para>Returns true only if this credential type has no scopes by default and requires 
```
Wrap true in c or in code
https://msdn.microsoft.com/en-us/library/vstudio/5ast78ax(v=vs.100).aspx

jtattermusch replied:
Done

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+
+        /// <summary>
+        /// <para>Returns true only if this credential type has no scopes by default and requires 
+        /// a call to createScoped before use.</para>
```
createScoped -> CreateScoped (capital C) and also add a comment link (see)

jtattermusch replied:
done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        /// 
+        /// <list type="number">
+        /// <item> 
+        /// <description>ComputeCredential has scopes built-in. Nothing additional is required.</description> 
```
Add links to ComputeCredential, UserCredential or whatever credential you mention

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+            get { return false; }
+        }
+
+        /// <summary>If the credential supports scopes, creates a copy with the specified scopes, otherwise return the same instance.</summary>
```
If the credential supports scopes, this method creates a copy with the specified scopes, otherwise it returns the same instance.

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+
+        #region ITokenAccess
+
+        TokenResponse ITokenAccess.Token
```
any reason why you use ITokenAcces.Token here? and not just Token?

jtattermusch replied:
I think the explicit implementation is better because it hides the members that won't be used in the most common scenario. This should help prevent user confusion or wrong usage of the API.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        /// <summary>Provides access to the underlying credential object</summary>
+        internal abstract object UnderlyingCredential { get; }
+
+        // We're explicitly implementing all the interfaces to only expose the members user actually
```
This is not a C# standard method comment... Please adjust

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        // We're explicitly implementing all the interfaces to only expose the members user actually
+        // needs to see. Because you cannot make explicit interface implementors abstract, they are redirecting
+        // to the following protected abstract members.
+        protected abstract void Initialize(ConfigurableHttpClient httpClient);
```
add a C# comment to this method

peleyal replied:
Not sure I'm following you. Why do we need these wrappers?

jtattermusch replied:
This note doesn't belong to the method itself but to the following three abstract members. In C#, it is illegal to explicitly implement an interface (which is what I want to do here) and leave those members as abstract. So I explicitly implement ITokenAccess.Token for example and redirect it to a protected abstract member (that the subclasses will implement).

This is an implementation detail that will go away once I address the TODO to make all credential classes implement the ITokenAccess interface. Let's leave it for now.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+
+        #region Factory methods
+
+        /// <summary>Creates a <c>GoogleCredential</c> wrapping a <c>ComputeCredential</c>.</summary>
```
links? here and there

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+        // TODO(jtattermush): Look into adjusting the API of ServiceAccountCredential, ComputeCredential
+        // and UserCredential so that they implement ITokenAccess. Then the boilerplate below will go away.
+
+        /// <summary>
```
By the way, a short summary can be fit in one line.

jtattermusch replied:
Done here and below.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,307 @@
+
+            public override GoogleCredential CreateScoped(IEnumerable<string> scopes)
+            {
+                // TODO(jtattermusch): make sure this is a good way to create a new service account credential
```
Is it still relevant?

jtattermusch replied:
Not anymore, thanks for noticing.

peleyal commented on Src/GoogleApis.Auth/OAuth2/ComputeCredential.cs:
```diff
@@ -99,5 +120,53 @@ public Initializer(string tokenUrl)
+            }
+            catch (HttpRequestException)
+            {
+                return false;
```
Log as well?

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,263 @@
+        /// <summary>Constructs a new default credential provider.</summary>
+        public DefaultCredentialProvider()
+        {
+            this.cachedCredentialTask = new Lazy<Task<GoogleCredential>>(CreateDefaultCredentialAsync);
```
remove the "this" keyword here.

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/GoogleCredential.cs:
```diff
@@ -0,0 +1,303 @@
+        // needs to see. Because you cannot make explicit interface implementors abstract, they are redirecting
+        // to the following protected abstract members.
+
+        protected abstract void Initialize(ConfigurableHttpClient httpClient);
```
Please add a comment to this method

jtattermusch replied:
Done.

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,268 @@
+        }
+
+        /// <summary>Creates a default credential from a stream that contains JSON credential data.</summary>
+        internal GoogleCredential CreateDefaultCredentialFromStream(Stream stream)
```
Should it be internal? Why?

jtattermusch replied:
GoogleCredential uses it from the GoogleCredential.FromStream method

peleyal commented on Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs:
```diff
@@ -0,0 +1,268 @@
+        }
+
+        /// <summary> 
+        /// Helper method to return well known credential file path. This file is created by 
```
A helper method?

jtattermusch replied:
Clarified the comment instead.

peleyal commented:
Great job Ian,
So sorry that it took forever, but I feel better with the code quality now.
thanks for addressing everything.

LGTM LGTM LGTM :+1:

jtattermusch commented:
Thanks for reviewing this, Eyal. 

I've rebased this pull request to clean up the history - now it has 2 commits, one of the the original work from @vsubramani  and the other one that has my work on top of it. Going to merge now.
